File: solutions/csharp/phone-number/PhoneNumber.cs
using System;
using System.Linq;

public class PhoneNumber
{
    public static string Clean(string phoneNumber)
    {
        // extract all digits from the given string
        var digits = phoneNumber.Where(Char.IsDigit).ToArray();

        // all NANP-numbers share the same country code
        if (digits.Length == 11 && digits[0] == '1')
        {
            digits = digits.Skip(1).ToArray();
        }

        // area and exchange codes only digits from 2 through 9
        if (digits.Length != 10
            || digits[0] is '0' or '1'
            || digits[3] is '0' or '1')
        {
            throw new ArgumentException("invalid phone number");
        }

        return new string(digits);
    }
}

File: solutions/elixir/phone-number/lib/phone_number.ex
defmodule PhoneNumber do
  @doc """
  Remove formatting from a phone number if the given number is valid. Return an error otherwise.
  """
  @spec clean(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def clean(raw) do
    digits = String.replace(raw, ~r/\D/, "")

    cond do
      String.match?(raw, ~r/[^\d-(). +]/) ->
        {:error, "must contain digits only"}

      String.length(digits) > 11 ->
        {:error, "must not be greater than 11 digits"}

      String.length(digits) < 10 ->
        {:error, "must not be fewer than 10 digits"}

      String.length(digits) == 11 and !String.starts_with?(digits, "1") ->
        {:error, "11 digits must start with 1"}

      String.at(digits, -10) == "0" ->
        {:error, "area code cannot start with zero"}

      String.at(digits, -10) == "1" ->
        {:error, "area code cannot start with one"}

      String.at(digits, -7) == "0" ->
        {:error, "exchange code cannot start with zero"}

      String.at(digits, -7) == "1" ->
        {:error, "exchange code cannot start with one"}

      true ->
        {:ok, String.slice(digits, -10, 10)}
    end
  end
end
